admin/services/update/update.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2019 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package update implements update API.
+package update
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/percona/pmm/admin/cli/flags"
+	dockerCmd "github.com/percona/pmm/admin/commands/pmm/server/docker"
+	"github.com/percona/pmm/admin/pkg/docker"
+	"github.com/percona/pmm/api/updatepb"
+)
+
+// Make sure this does not include any directories, just a basename.
+const logsFileNamePattern = "upgrade.*.log"
+
+type Server struct {
+	// Context coming from cli commands. When cancelled, the command has been cancelled.
+	cliCtx             context.Context
+	dockerFn           dockerFunctions
+	dockerImage        string
+	gRPCMessageMaxSize uint32
+	updateInProgress   map[string]struct{}
+	updateMu           sync.RWMutex
+
+	updatepb.UnimplementedUpdateServer
+}
+
+// New returns new instance of Server.
+func New(ctx context.Context, dockerImage string, gRPCMessageMaxSize uint32) (*Server, error) {
+	d, err := docker.New(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Server{
+		cliCtx:             ctx,
+		dockerFn:           d,
+		dockerImage:        dockerImage,
+		gRPCMessageMaxSize: gRPCMessageMaxSize,
+		updateInProgress:   make(map[string]struct{}, 1),
+	}, nil
+}
+
+// StartUpdate starts PMM Server update.
+func (s *Server) StartUpdate(ctx context.Context, req *updatepb.StartUpdateRequest) (*updatepb.StartUpdateResponse, error) {
+	containerID := req.Hostname
```

Comment:
The value is provided by PMM Server. The server does not have a concept of a container ID, it only knows its hostname.
The hostname is the container ID in Docker, unless you override it.
The assignment thus makes sense. But I guess we could add detection based on hostname in case it's not the container ID. Not sure how many customers set a custom hostname though. I expect very little.

Reply:
But since it's API of `pmm-server-updater` it should be `containerID` here.

Reply:
If we rename it to `containerID`, PMM Server has no way to provide the ID of the container since it does not know the ID.
Maybe we keep it a `hostname` and I implement the extra logic as described above?

Reply:
API should be implemented based on server needs, not a client